pkg/util/servicediscovery/ring_test.go

```diff
@@ -14,42 +14,137 @@ import (
 	"github.com/grafana/dskit/ring"
 	"github.com/grafana/dskit/services"
 	"github.com/grafana/dskit/test"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
-func TestRingServiceDiscovery(t *testing.T) {
+func TestRingServiceDiscovery_WithoutMaxUsedInstances(t *testing.T) {
 	const (
 		ringKey         = "test"
 		ringCheckPeriod = 100 * time.Millisecond // Check very frequently to speed up the test.
 	)
 
-	var (
-		ctx    = context.Background()
-		ringOp = ring.NewOp([]ring.InstanceState{ring.ACTIVE}, nil)
+	// Use an in-memory KV store.
+	inmem, closer := consul.NewInMemoryClient(ring.GetCodec(), log.NewNopLogger(), nil)
+	t.Cleanup(func() { _ = closer.Close() })
+
+	// Create a ring client.
+	ringCfg := ring.Config{HeartbeatTimeout: time.Minute, ReplicationFactor: 1}
+	ringClient, err := ring.NewWithStoreClientAndStrategy(ringCfg, "test", ringKey, inmem, ring.NewDefaultReplicationStrategy(), nil, log.NewNopLogger())
+	require.NoError(t, err)
+
+	// Create an empty ring.
+	ctx := context.Background()
+	require.NoError(t, inmem.CAS(ctx, ringKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		return ring.NewDesc(), true, nil
+	}))
+
+	// Mock a receiver to keep track of all notified addresses.
+	receiver := newNotificationsReceiverMock()
+
+	sd := NewRing(ringClient, ringCheckPeriod, 0, receiver)
+
+	// Start the service discovery.
+	require.NoError(t, services.StartAndAwaitRunning(ctx, sd))
+	t.Cleanup(func() {
+		require.NoError(t, services.StopAndAwaitTerminated(ctx, sd))
+	})
+
+	// Wait some time, we expect no address notified because the ring is empty.
+	time.Sleep(time.Second)
+	require.Empty(t, receiver.getDiscoveredInstances())
+
+	// Register some instances.
+	require.NoError(t, inmem.CAS(ctx, ringKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		desc := in.(*ring.Desc)
+		desc.AddIngester("instance-1", "instance-1", "", nil, ring.ACTIVE, time.Now())
+		desc.AddIngester("instance-2", "instance-2", "", nil, ring.PENDING, time.Now())
+		desc.AddIngester("instance-3", "instance-3", "", nil, ring.JOINING, time.Now())
+		desc.AddIngester("instance-4", "instance-4", "", nil, ring.LEAVING, time.Now())
+		return desc, true, nil
```

Comment:
I would suggest to use address that's different from instance names (also in the next test). This may hide bug when ID is reported instead of address. (We should include port in the address to make it look like a real ring).

Reply:
Good idea!